pkg/controller/volume/pvcprotection/pvc_protection_controller.go

```diff
@@ -418,3 +440,62 @@ func (c *Controller) enqueuePVCs(pod *v1.Pod, deleted bool) {
 		}
 	}
 }
+
+// This function checks if the PV correspodning to the PVC is still attached to
+// some node.
+func (c *Controller) isPVCAttachedToNodes(pvc *v1.PersistentVolumeClaim) (bool, error) {
+	volName := pvc.Spec.VolumeName
+	// isPVCAttachedToNodes is called only when PVC is a delete candidate. In the
+	// scenario where a PVC is already marked for deletion and the volume name is
+	// not stamped on the PVC, skip the nodes attachment check.
+	if volName == "" {
+		klog.Infof("Failed to fetch PV name for the PVC %s/%s which is marked for deletion, skip node attached check for PV", pvc.Namespace, pvc.Name)
+		return false, nil
+	}
+
+	// Check node status in the informer cache, if error encountered or volume not found, do not return immediately, try the API server for consistency.
+	if nodeName, ok := pvc.Annotations[AnnSelectedNode]; ok {
+		klog.V(3).Infof("Checking for volume %s in informer cache for selected node %s", volName, nodeName)
+		selectedNode, err := c.nodeLister.Get(nodeName)
+		nodes := []*v1.Node{selectedNode}
+		if err != nil {
+			klog.V(3).Infof("unexpected error getting target node %q from node informer for claim %s/%s: %v", nodeName, pvc.Namespace, pvc.Name, err)
+		} else if c.isVolumeInUse(volName, nodes) {
+			return true, nil
+		}
+	} else {
+		klog.V(3).Infof("Checking for volume %s in node informer cache", volName)
+		nodes, err := c.nodeLister.List(labels.Everything())
+		if err != nil {
+			klog.V(3).Infof("unexpected error listing nodes from node informer: %v", err)
+		} else if c.isVolumeInUse(volName, nodes) {
+			return true, nil
+		}
+	}
+
+	// If we reach here, we could not determine if the volume is attached to any node.
+	// Fetch node list from API server to check node.Status.VolumesInUse.
+	nodes, err := c.client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
```

Comment:
I think normally if nodeLister.List failed, we return the error, and controller will try again. We want to avoid calling API server directly to avoid constant API calls in case of failure.

Reply:
Ack good point.